src/clients/c++/library/request_grpc.cc

```diff
@@ -944,31 +927,32 @@ InferGrpcContextImpl::AsyncTransfer()
       if (this->exiting_) {
         return true;
       }
-      // wake up if at least one request is not ready
-      for (auto& ongoing_async_request : this->ongoing_async_requests_) {
-        if (std::static_pointer_cast<GrpcRequestImpl>(
-                ongoing_async_request.second)
-                ->IsReady() == false) {
-          return true;
-        }
+      // wake up if an async request has been generated
+      if (!this->ongoing_async_requests_.empty()) {
+        return true;
       }
       return false;
     });
     lock.unlock();
     // GRPC async APIs are thread-safe https://github.com/grpc/grpc/issues/4486
     if (!exiting_) {
-      std::shared_ptr<Request> request_with_callback;
+      std::shared_ptr<Request> request;
       size_t got;
       bool ok = true;
       bool status = async_request_completion_queue_.Next((void**)(&got), &ok);
+      if (!ok) {
+        // The lock would prevent the jumbling up of the error messages.
+        // Additionally, doesn't impact performance as it is outside the
+        // execution path.
+        std::lock_guard<std::mutex> lock(mutex_);
+        fprintf(stderr, "Unexpected not ok on client side.");
+      }
+      if (!status) {
+        std::lock_guard<std::mutex> lock(mutex_);
+        fprintf(stderr, "Completion queue is closed.");
```

Comment:
We are just logging error here, do we need to handle them? The part below will still be executed on error...

Reply:
Fixed